### PR TITLE
User disconnected

### PR DESCRIPTION
**Pull request by Claudio5**
This PR implements:
- When a user loses connection in any state of the online game (i.e. WaitingPageActivity, DrawingOnline(Activities), VotingPageActivity and RankingFragment) a popUp shows up and redirects user to loginPage.

- A user that leaves or disconnects will automatically finish in 5th position and thus losing automatically trophies.

- If users leave before VotingActivity a disconnected message will be displayed in VotingActivity and in RankingFrangment. (If user leaves in voting activity the disconnected will only appear in RankingFragment). It is not possible to vote for disconnected players.

- Disconnected users or users that leave before the game ends will have their stars/trophies/leagues updated by the server.

N.B. : For now quitting the app during a game is considered as leaving the game.

### CODE REVIEW

File: app/src/main/java/ch/epfl/sweng/SDP/game/RankingFragment.java [hunk 1]
```diff
@@ -233,13 +234,25 @@ public View getView(int position, View convertView, @NonNull ViewGroup parent) {
             convertView = LayoutInflater.from(getContext())
                     .inflate(R.layout.ranking_item, parent, false);
 
-            setTypeFace(TypefaceLibrary.getTypeMuro(), convertView.findViewById(R.id.playerName),
-                    convertView.findViewById(R.id.starsWon),
-                    convertView.findViewById(R.id.trophiesWon));
+            AssetManager assets = getActivity().getAssets();
+            Typeface typeMuro = Typeface.createFromAsset(assets, "fonts/Muro.otf");
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/game/RankingFragment.java]: You should use TypefaceLibrary here

File: app/src/main/java/ch/epfl/sweng/SDP/game/VotingPageActivity.java [hunk 2]
```diff
@@ -189,16 +222,15 @@ public void onCancelled(@NonNull DatabaseError databaseError) {
                 throw databaseError.toException();
             }
         });
+
+        Database.getReference(TOP_ROOM_NODE_ID + "." + roomID + ".onlineStatus."
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/game/VotingPageActivity.java]: This code is used several times, could be a good idea extracting it in a method

File: app/src/main/java/ch/epfl/sweng/SDP/game/VotingPageActivity.java [hunk 3]
```diff
@@ -121,32 +140,46 @@ protected void onCreate(Bundle savedInstanceState) {
         setContentView(R.layout.activity_voting_page);
         overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left);
 
+        NetworkStateReceiverListener networkStateReceiverListener =
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/game/VotingPageActivity.java]: Maybe it’s better to extract the code handling the connectivity in a separated method with a meaningful name

Comment (Gondolav) [reply app/src/main/java/ch/epfl/sweng/SDP/game/VotingPageActivity.java]: In fact, it is used in several classes. Is it possible to wrap it directly in an another utility class, to avoid code duplication?

File: app/src/main/java/ch/epfl/sweng/SDP/game/drawing/DrawingOnline.java [hunk 4]
```diff
@@ -113,6 +119,16 @@ public int getLayoutId() {
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+
+        Typeface typeMuro = Typeface.createFromAsset(getAssets(), "fonts/Muro.otf");
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/game/drawing/DrawingOnline.java]: Here too TypefaceLibrary should be used

File: app/src/main/java/ch/epfl/sweng/SDP/game/drawing/DrawingOnline.java [hunk 5]
```diff
@@ -128,20 +144,17 @@ protected void onCreate(Bundle savedInstanceState) {
         timerRef.addValueEventListener(listenerTimer);
         stateRef = Database.getReference(TOP_ROOM_NODE_ID + "." + roomId + ".state");
         stateRef.addValueEventListener(listenerState);
+
+        Database.getReference(TOP_ROOM_NODE_ID + "." + roomId + ".onlineStatus."
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/game/drawing/DrawingOnline.java]: Since it is used here too, it could be wrapped in a separated class (maybe the same wrapping the connectivity code a few lines above)

File: app/src/main/java/ch/epfl/sweng/SDP/utils/RankingUtils.java [hunk 6]
```diff
@@ -43,18 +49,23 @@ private RankingUtils() {
      */
     public static Integer[] generatePositionsFromRanking(Integer[] ranking) {
         Integer[] positions = new Integer[ranking.length];
-        positions[0] = 1;
+
+        positions[0] = ranking[0] >= 0 ? 1 : 5;
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/RankingUtils.java]: The magic numbers should be replaced by constants as in the other method

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Utility class for checking device's connectivity.
+ */
+public class NetworkStateReceiver extends BroadcastReceiver {
+
+    protected Set<NetworkStateReceiverListener> listeners;
+    protected Boolean connected;
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Can these fields be private?

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiverListener.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package ch.epfl.sweng.SDP.utils.network;
+
+public interface NetworkStateReceiverListener {
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiverListener.java]: You should add javadocs for this interface

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStatusHandler.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import android.app.Dialog;
+import android.content.Context;
+import android.content.Intent;
+import android.view.View;
+
+import ch.epfl.sweng.SDP.MainActivity;
+import ch.epfl.sweng.SDP.R;
+import ch.epfl.sweng.SDP.auth.Account;
+
+public class NetworkStatusHandler implements NetworkStateReceiverListener {
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStatusHandler.java]: You missed a javadoc here!

File: functions/index.js [hunk 10]
```diff
@@ -154,6 +156,16 @@ exports.joinGame2 = functions.https.onCall((data, context) => {
   });
 });
 
+function getLeagueFromTrophies(trophies) {
+  if(trophies < 100 && trophies >= 0) {
```
Comment (Gondolav) [inline functions/index.js]: Just for clarity, the two checks should be inverted (sorry for that haha)

File: functions/index.js [hunk 11]
```diff
@@ -255,39 +267,103 @@ function startTimer(time, roomID, prevState, newState) {
...
-      admin.database().ref("users/" + userID + "/trophies").transaction((currentValue) => {
-          return currentValue + getRankFromPosition(position);
+      let newTrophies;
+      let totalMatches;
+
+      return admin.database().ref("users/" + userID + "/trophies").transaction((currentValue) => {
+        newTrophies = currentValue + minRank;
+        return newTrophies;
+      })
+      .then(() => {
```
Comment (Gondolav) [inline functions/index.js]: Is it possible to do all the updates in parallel in this case? Just asking out of curiosity, you don’t need to actually do it haha

Comment (Claudio5) [reply functions/index.js]: Yes you can and that's what I did at the beginning but for some reason there was a weird bug with this. But for example the league update needs to wait for the trophies update so they need to be sequential as well as the change average which needs to wait for the increase match. So we could potentially split in two.

File: app/src/main/java/ch/epfl/sweng/SDP/utils/RankingUtils.java [hunk 12]
```diff
@@ -21,12 +22,17 @@ private RankingUtils() {
      */
     public static Integer[] generateTrophiesFromRanking(Integer[] ranking) {
         Integer[] trophies = new Integer[ranking.length];
-        trophies[0] = MAX_RANK;
+
```
Comment (FredrikBixo) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/RankingUtils.java]: Why did you change this method? (Substitute algorithm)

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Utility class for checking device's connectivity.
+ */
+public class NetworkStateReceiver extends BroadcastReceiver {
```
Comment (FredrikBixo) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Good use of the observer pattern

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        if (connected) {
+            listener.networkAvailable();
+        } else {
+            listener.networkUnavailable();
+        }
+    }
+
+    public void addListener(NetworkStateReceiverListener listener) {
```
Comment (FredrikBixo) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Add javadoc?

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        listeners.add(listener);
+        notifyState(listener);
+    }
+
+    public void removeListener(NetworkStateReceiverListener listener) {
```
Comment (FredrikBixo) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Same here

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        listeners.remove(listener);
+    }
+
+    public Set<NetworkStateReceiverListener> getListeners() {
```
Comment (FredrikBixo) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Same here

File: functions/index.js [hunk 17]
```diff
@@ -248,46 +260,93 @@ exports.onUsersChange = functions.database.ref(parentRoomID + "{roomID}/users").
...
+    snapshot.forEach((child) => {
+      console.log(child.val());
+      if(child.val() === 0) {
+        admin.database().ref(parentRoomID + roomID + "/ranking/" + child.key).set(-1);
+      }
+    });
+  }).then(() => {return createNode(roomID, "onlineStatus", 0)});
+}
+
 exports.onStateUpdate = functions.database.ref(parentRoomID + "{roomID}/state").onWrite((change, context) => {
```
Comment (FredrikBixo) [inline functions/index.js]: This code looks like a mess. Consider extract method here :)

File: app/src/androidTest/java/ch/epfl/sweng/SDP/game/VotingPageActivityTest.java [hunk 18]
```diff
@@ -157,20 +163,32 @@ public void startHomeActivityStartsHomeActivity() {
...
         intended(hasComponent(HomeActivity.class.getName()));
         Intents.release();
+        Database.getReference("realRooms.0123457890.users." + USER_ID).setValue(USER_ID);
+        Database.getReference("realRooms.0123457890.ranking." + USER_ID).setValue(0);
+        SystemClock.sleep(2000);
     }
 
     @Test
-    public void testState5Change() {
+    public void testState6Change() {
```
[the remark this answers is not in the document]
Comment (Claudio5) [reply app/src/androidTest/java/ch/epfl/sweng/SDP/game/VotingPageActivityTest.java]: In our Firebase database we maintain states from 0 to 6. That's why I am using these numbers for the signature of the function.

File: app/src/androidTest/java/ch/epfl/sweng/SDP/game/WaitingPageActivityTest.java [hunk 19]
```diff
@@ -48,6 +57,9 @@
 
     private DataSnapshot dataSnapshotMock;
     private DatabaseError databaseErrorMock;
+    private Intent intentMock;
+    private ConnectivityManager cmMock;
+    private Context spyContext;
```
[the remark this answers is not in the document]
Comment (Claudio5) [reply app/src/androidTest/java/ch/epfl/sweng/SDP/game/WaitingPageActivityTest.java]: I agree from cmMock maybe. But not for spyContext, I mean it's not a mock but we spy directly from the instance of the context. So instead of calling it mockContext it's just spyContext. We keep the same syntax as the above member variables.

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java (hunk 13 above)
Comment (Claudio5) [reply app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Thanks

File: functions/index.js (hunk 17 above)
Comment (Claudio5) [reply functions/index.js]: Javascript is a mess in general ;)

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java (hunk 7 above)
Comment (Claudio5) [reply app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Yes !

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java (hunk 16 above)
Comment (Claudio5) [reply app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: This is a getter no need for javadoc from my point of view

File: app/src/androidTest/java/ch/epfl/sweng/SDP/game/WaitingPageActivityTest.java [hunk 20]
```diff
@@ -317,4 +332,52 @@ public void testVoteAndRemoveVote() {
         intended(hasComponent(HomeActivity.class.getName()));
         Intents.release();
     }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (Claudio5) [reply app/src/androidTest/java/ch/epfl/sweng/SDP/game/WaitingPageActivityTest.java]: It checks that the pop up is displayed `onView(withId(R.id.okDisconnectedButton)).check(matches(isDisplayed()));`

File: app/src/main/java/ch/epfl/sweng/SDP/game/RankingFragment.java [hunk 21]
```diff
@@ -132,7 +133,7 @@ public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                 int trophiesForUser = trophies[usernames.indexOf(account.getUsername())];
 
                 Boolean won = usernames.get(0).equals(account.getUsername());
-                updateUserStats(rankForUser, trophiesForUser, won);
+                updateUserStats(Math.max(rankForUser, 0), trophiesForUser, won);
```
[the remark this answers is not in the document]
Comment (Claudio5) [reply app/src/main/java/ch/epfl/sweng/SDP/game/RankingFragment.java]: Ranking cannot be negative because ranking is the number of stars you earn. But in my implementation I set negative ranking to players that are disconnected in order to treat them differently. But in the end I do not want to show a negative ranking thus the max function

File: app/src/main/java/ch/epfl/sweng/SDP/game/RankingFragment.java [hunk 22]
```diff
@@ -233,13 +234,25 @@ public View getView(int position, View convertView, @NonNull ViewGroup parent) {
...
+                                convertView.findViewById(R.id.starsWon),
+                                convertView.findViewById(R.id.trophiesWon),
+                                convertView.findViewById(R.id.disconnectedRanking));
 
             // Update image
-            ImageView imageview = convertView.findViewById(R.id.drawing);
-            imageview.setImageBitmap(drawings[getIndexForUserName(players[position])]);
+            ImageView imageView = convertView.findViewById(R.id.drawing);
+            TextView disconnectedTextView = convertView.findViewById(R.id.disconnectedRanking);
+            if (rankings[position] >= 0) {
```
[the remark this answers is not in the document]
Comment (Claudio5) [reply app/src/main/java/ch/epfl/sweng/SDP/game/RankingFragment.java]: Sure, in short if players are disconnected it does not show their drawings but a disconnected textview instead. 
As explained above I set ranking to negative value to know if players is disconnected or left. Thus by just checking the ranking I know whether I display a drawing or textview.

File: app/src/main/java/ch/epfl/sweng/SDP/game/VotingPageActivity.java [hunk 23]
```diff
@@ -214,12 +246,17 @@ public void startHomeActivity(View view) {
         // Remove the drawings from Firebase Storage
         for (String id : drawingsIds) {
```
[the remark this answers is not in the document]
Comment (Claudio5) [reply app/src/main/java/ch/epfl/sweng/SDP/game/VotingPageActivity.java]: I am not sure to understand, which method ?

File: app/src/main/java/ch/epfl/sweng/SDP/MainActivity.java [hunk 24]
```diff
@@ -34,7 +39,8 @@ protected void onCreate(Bundle savedInstanceState) {
         FirebaseApp.initializeApp(this);
         FirebaseAuth auth = FirebaseAuth.getInstance();
 
-        if (auth.getCurrentUser() != null) {
+        // Go to the home if the user has already logged in and created an account
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/MainActivity.java]: You can remove this comment, I moved it below

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+import android.content.IntentFilter;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+import android.support.annotation.VisibleForTesting;
+
+import ch.epfl.sweng.SDP.firebase.Database;
+import ch.epfl.sweng.SDP.game.WaitingPageActivity;
+
+/**
+ * Connectivity wrapper that registers the the network receiver and unregisters it.
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java]: Twice the the ;)

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+ */
+public final class ConnectivityWrapper {
+
+    private static final String TOP_ROOM_NODE_ID = "realRooms";
+    private static NetworkStateReceiver networkStateReceiver = null;
+
+    private ConnectivityWrapper() {}
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java]: Inside the constructor you should check for instance != null, and if it’s the case, an IllegalStateException should be thrown

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+    private static NetworkStateReceiver getInstanceNetwork() {
+        if (networkStateReceiver == null) {
+            networkStateReceiver = new NetworkStateReceiver();
+        }
+
+        return networkStateReceiver;
+    }
+
+    /**
+     * Create a network receiver and sets the corresponding listener to it.
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java]: Creates

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+        NetworkStateReceiverListener networkStateReceiverListener =
+                new NetworkStatusHandler(context);
+
+        networkStateReceiver.addListener(networkStateReceiverListener);
+        context.registerReceiver(networkStateReceiver,
+                new IntentFilter(android.net.ConnectivityManager.CONNECTIVITY_ACTION));
+    }
+
+    /**
+     * Unregister the listener and deletes the instance of the receiver.
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java]: Unregisters

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+     */
+    public static void unregisterNetworkReceiver(Context context) {
+        if(networkStateReceiver != null) {
+            context.unregisterReceiver(networkStateReceiver);
+            networkStateReceiver = null;
+        }
+    }
+
+    /**
+     * Set the online flag to 1 when in-game. Allows to notify database that the player is still
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/ConnectivityWrapper.java]: Sets

File: app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+     *
+     * @param listener Listener to be added
+     */
+    public void addListener(NetworkStateReceiverListener listener) {
+        listeners.add(listener);
+        notifyState(listener);
+    }
+
+    /**
+     * Remove from the listener sets the given listener.
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/network/NetworkStateReceiver.java]: Removes

File: app/src/main/java/ch/epfl/sweng/SDP/utils/RankingUtils.java (hunk 12 above)
Comment (Claudio5) [reply app/src/main/java/ch/epfl/sweng/SDP/utils/RankingUtils.java]: The method was improved
